OrbitQt/ConnectionConfiguration.h

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_QT_CONNECTION_CONFIGURATION_H_
+#define ORBIT_QT_CONNECTION_CONFIGURATION_H_
+
+#include <memory>
+#include <utility>
+
+#include "DeploymentConfigurations.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitClientData/ProcessData.h"
+#include "OrbitClientServices/ProcessManager.h"
+#include "OrbitGgp/Instance.h"
+#include "OrbitSsh/Context.h"
+#include "OrbitSsh/Credentials.h"
+#include "grpcpp/channel.h"
+#include "servicedeploymanager.h"
+
+namespace orbit_qt {
+
+struct StadiaConnection {
+ public:
+  explicit StadiaConnection(const OrbitSsh::Context* ssh_context,
+                            const ServiceDeployManager::GrpcPort& grpc_port,
+                            const DeploymentConfiguration* deployment_configuration)
+      : ssh_context(ssh_context),
+        grpc_port(grpc_port),
+        deployment_configuration(deployment_configuration) {
+    CHECK(ssh_context != nullptr);
+    CHECK(deployment_configuration != nullptr);
+  }
+  ~StadiaConnection() {
+    if (process_manager != nullptr) process_manager->Shutdown();
+  }
+  void CreateServiceDeployManager(OrbitSsh::Credentials credentials) {
+    CHECK(service_deploy_manager == nullptr);
+    service_deploy_manager = std::make_unique<ServiceDeployManager>(
+        deployment_configuration, ssh_context, std::move(credentials), grpc_port);
+  }
+
```

Comment:
That doesn't feel right. This gives that struct invariants which it shouldn't have by definition.

Also this class has state that is not really "runtime-state". It's just mimicking two "programming-stages" (without and with ServiceDeployManager). Ideally we represent that in static types instead of dynamically handling
two branches.

I think you should split that up into two classes/struct. One class that holds all resources that are needed
for constructing the ServiceDeployManager and one class that holds all the elements which are needed
to be given to the main window. And then this function can be a free function taking an r-value-reference of
the first class and those credentials and returning a value of the second class.

Reply:
I used classes now, instead of structs. The classes in this PR (now in Connections.h) are actually holding less data than before, because some of it went into TargetConfiguration.h which will be part of #1526.